packages/flutter/lib/src/semantics/semantics.dart

```diff
@@ -354,6 +359,101 @@ sealed class _DebugSemanticsRoleChecks {
     }
     return null;
   }
+
+  static bool _isLandmarkRole(SemanticsNode node) =>
+      node.role == SemanticsRole.complementary ||
+      node.role == SemanticsRole.contentInfo ||
+      node.role == SemanticsRole.main ||
+      node.role == SemanticsRole.navigation ||
+      node.role == SemanticsRole.region;
+
+  static bool _isSameRoleExisted(SemanticsNode semanticsNode) {
+    final Map<int, SemanticsNode> treeNodes = semanticsNode.owner!._nodes;
+    int sameRoleCount = 0;
+    for (final int id in treeNodes.keys) {
+      if (treeNodes[id]?.role == semanticsNode.role) {
+        sameRoleCount++;
+        if (sameRoleCount > 1) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  static FlutterError? _semanticsComplementary(SemanticsNode node) {
+    SemanticsNode? currentNode = node.parent;
+    while (currentNode != null) {
+      if (_isLandmarkRole(currentNode)) {
```

Comment:
when checking the property or flag, we should check on semanticsData returned from getSemanticsData(). This will make sure we take mergeSemantics into account